### ISSUE TITLE
recommended settings for discovering multiple rfduinos

### ISSUE THREAD

**Issue by batman11111** [don/cordova-plugin-rfduino#23]
Hi Don, what settings do you recommend to discover multiple rfduinos? I'm trying to discover 7 of them, I have the advertising interval set to 500ms and the discovery time set to 7 seconds. My android app discovers 4 of them every time and the other 3 are hit and miss (sometimes it finds them in the 7 seconds, sometimes it doesn't) is this a limitation with the plugin or do you have  any ideas on how to discover all 7 every time?

**Comment by don** [comment]
It seems like you should be able to discover all of your RFduinos with those settings, but I haven't tried that many at once.. The plugin just passes data though to the iOS or Android native APIs for discovery. Are you seeing the problem on iOS or Android? Are any of the the missing RFduinos connected to another device? Are the same 3 devices always the problem or is it random devices? Are any devices connected when you have missing ones? Connected devices won't advertise.  Other than that, I'd decrease the advertising interval if possible and perhaps increase the discovery to 10 seconds.

**Comment by batman11111** [comment]
I'm seeing it on Android right now, haven't tried it on iOS yet.  None of the devices are not connected to anything.  I have the devices named 1 -7.  1-5 Always show up within a few seconds.  6 and 7 often take up to 30 seconds to show up.  The code on all the rfduinos is exactly the same (except the device name).  I've downloaded a 3rd party BLE scanner and it gives me the same results(which tells me it probably doesn't have anything to do with your plugin).  Any ideas on what could be happening though?

**Comment by batman11111** [comment]
I'm assuming rfduinos 6 and 7 may just be bad?  I've replaced them with 2 new rfduinos and it discovers all of them within a few seconds.  (Its strange because they work, they just don't always get discovered as quick as the rest, I don't understand why)  Anyways, another question:

Can you "manually" connect to an rfdunio.   For example if you know the uuid can you just type rfduino.connect(aa:bb:cc:dd:ee:ff, succcess, fail)  without running  a discovery?  I tried running a discovery to discover all 7 uuid and then I saved them to local storage in an array.  If I re-run the app without running rfduino.discover and just try connecting with rfduino.connect(uuidArray[value], success, fail; and it fails every time. I have verified it is indeed using a valid uuid by putting an alert(uuidArray[value]); right before it. Can you explain briefly why it fails, or should it work and I just have an error in my code somewhere? Thanks for all you help.

**Comment by don** [comment]
Maybe the bad/slow RFduinos need a firmware update? If you have a bluetooth sniffer, you could see how often the devices are advertising. I use [Adafruit's Bluefruit LE Sniffer](https://www.adafruit.com/products/2269).

You can manually connect to a RFduino by entering the MAC address, but you need to run scan first, since my code expects to have a peripheral found from scanning. It's possible the plugin could be adjusted to allow direct connections, but you'd need to dig into the Android code to ensure this was possible.

**Comment by batman11111** [comment]
Thanks for the info, just ordered the bluetooth sniffer.  I also tried 7 completely different rfduinos, and still had similar results (one would only discover about 80% of the time while the rest were fine). I opened a case with rfduino to see if they can help track down the issue.  Feel free to close this on github unless you want me to update it with any info I get from rfduino.  Thanks again for all your help, definitely appreciate it.

**Comment by batman11111** [comment]
For anyone else experiencing this issue, I believe I have found a workaround. 

It seems with more then 3 rfduinos there will be collisions and a random device (or two) will become very  unstable.  

Anywhere from 15 seconds to over 2 minutes at a time the device will "freeze" and not advertise or allow connections.  This continues to happen frequently and randomly until the device is reset.  

My solution so far has been to base my sketch off of the rfduino periodicAdvertisment example using randomly generated numbers -to help avoid collisions.  I also added a watchdog timer to my sketch.  This automatically interrupts and resets the device if it happens to "freeze".  Since making these changes, all devices seem to be running as they should.

I'm still working with rfdigital for a fix.  I'll post any new information I find.
